docs/extra/optimize_function.rst

```diff
@@ -0,0 +1,297 @@
+  ...     res = method(to_optimize, start, lo, hi)
+  ...     if res[0]:
+  ...         tbl.add_row([method.__name__, res[2], res[1][0], res[1][1]])
+  ...     else:
+  ...         print(f"Failed {method.__name__}: {res[3]}")
+  ...
+  Failed lmdif: improper input parameters
+  >>> print(tbl)
+      method           stat0                 x                  y
+    ---------- --------------------- ------------------ ------------------
+         minim  1.73832000295664e-11 1.9999999630648686  4.000644742396361
+    neldermead 4.932797434506576e-11 2.0000007686008545 3.9994204823962214
+    montecarlo 6.629393582604866e-12 1.9999997816942297  3.999564165982678
+
+In this particular case :py:func:`~sherpa.optmethods.optfcts.lmdif`
+failed, and this is because the ``to_optimize`` function returned None
+as its second argument. For the other cases we can see that they all
+found the minimum location.
+
+In order to use :py:func:`~sherpa.optmethods.optfcts.lmdif` we need
+the per-parameter statistic, which we can get with a small tweak::
+
+  def to_optimize2(args):
+      x = args[0]
+      y = args[1]
+      pa = (a - x) * (a - b)
+      pb = b * (y - x * x) * (y - x * x)
+      stat = pa * pa + pb * pb
+      return stat, [pa, pb]
+
+This lets us use `lmdif`::
+
+  >>> res2 = optfcts.lmdif(to_optimize2, start, lo, hi)
+  >>> res2[0]
+  True
+  >>> res2[1]
+  [2.        4.0000846]
+
+The `lmdif` optimizer is one of those that returns a number of
+elements in the "extra" dictionary::
+
+  >>> res2[4]
+  {'info': 0, 'nfev': 108, 'covar': array([[ 1.56250000e-02, -1.04361202e-01],
+         [-1.04361202e-01,  4.15098457e+03]]), 'num_parallel_map': 0}
+
+Optimizing a model
+------------------
+
+We can re-create the :ref:`Fitting the NormGauss1D and Lorentz1D models <fit_peaked_data_normgauss1d_lorentz1d>`
+section - at least for the :py:class:`~sherpa.models.basic.NormGauss1D` fit.
+
+The normalized gaussian model can be expressed as
+
+::
+
+  def ngauss(x, ampl, pos, fwhm):
+      term = 4 * np.log(2)
+      numerator = ampl * np.exp(-term * (x - pos) * (x - pos) / (fwhm * fwhm))
+      denominator = np.sqrt(np.pi / term) * fwhm
+      return numerator / denominator
+
+and the data we are going to fit is read from a file:
+
+  >>> d = np.loadtxt('test_peak.dat')
+  >>> x = d[:, 0]
+  >>> y = d[:, 1]
+
+In this case the "callback" routine we want to give the minimizer a
+routine which will apply a least-squares minimizer to compare the
+model to the data (the `x` and `y` values for the independent and
+dependent axes respectively).
+
+::
+
+  def cb(pars):
+      model = ngauss(x, pars[0], pars[1], pars[2])
+      delta = model - y
+      statval = (delta * delta).sum()
+      # Keep a record of the parameters we've visited
+      store.add_row([statval] + list(pars))
+      return statval, delta
+
+.. note::
+   This function would normally be written as a closure to ensure
```

Comment:
I don't know what a "closure" is, I would use a callable class that keep that info as a class attribute. But either way, the point is to point out that that's a simply example, so whatever works here.
Or maybe a callable class is just one example to implement a  "closure"? I need to google and learn that term!

Reply:
A closure "encloses" the symbols it uses and I have seen it used in python discussions - even Guido has used it ...

https://www.python.org/dev/peps/pep-3104/#id47

Reply:
I've rewritten this to say something "a class or a closure"